common/rpc/rpc.go

```diff
@@ -0,0 +1,239 @@
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package rpc
+
+import (
+	"crypto/tls"
+	"fmt"
+	"net"
+	"sync"
+
+	"github.com/uber/tchannel-go"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+
+	"github.com/temporalio/temporal/common/log"
+	"github.com/temporalio/temporal/common/log/tag"
+	"github.com/temporalio/temporal/common/rpc/encryption"
+	config "github.com/temporalio/temporal/common/service/config"
+)
+
+// RPCFactory is an implementation of service.RPCFactory interface
+type RPCFactory struct {
+	config      *config.RPC
+	serviceName string
+	logger      log.Logger
+
+	sync.Mutex
+	grpcListener   net.Listener
+	ringpopChannel *tchannel.Channel
+	tlsFactory     encryption.TLSFactory
+}
+
+// NewFactory builds a new RPCFactory
+// conforming to the underlying configuration
+func NewFactory(cfg *config.RPC, sName string, logger log.Logger, serverCfg *config.Global) (*RPCFactory, error) {
+	tlsFactory, err := encryption.NewTLSFactoryFromConfig(
+		&serverCfg.TLS,
+		getBroadcastAddressFromConfig(serverCfg, cfg, logger))
+
+	if err != nil {
+		return nil, err
+	}
+	return newFactory(cfg, sName, logger, tlsFactory)
+}
+
+func newFactory(cfg *config.RPC, sName string, logger log.Logger, frontendTls encryption.TLSFactory) (*RPCFactory, error) {
+	factory := &RPCFactory{config: cfg, serviceName: sName, logger: logger, tlsFactory: frontendTls}
+	return factory, nil
+}
+
+func (d *RPCFactory) GetFrontendGRPCServerOptions() ([]grpc.ServerOption, error) {
+	opts := make([]grpc.ServerOption, 0, 1)
+
+	if d.tlsFactory != nil {
+		serverConfig, err := d.tlsFactory.GenerateFrontendServerConfig()
+		if err != nil {
+			return nil, err
+		}
+		opts = append(opts, grpc.Creds(credentials.NewTLS(serverConfig)))
+	}
+
+	return opts, nil
+}
+
+func (d *RPCFactory) GetFrontendClientTlsConfig() (*tls.Config, error) {
+	if d.tlsFactory != nil {
+		return d.tlsFactory.GenerateFrontendClientConfig()
+	}
+
+	return nil, nil
+}
+
+func (d *RPCFactory) GetInternodeGRPCServerOptions() ([]grpc.ServerOption, error) {
+	opts := make([]grpc.ServerOption, 0, 1)
+
+	if d.tlsFactory != nil {
+		serverConfig, err := d.tlsFactory.GenerateInternodeServerConfig()
+		if err != nil {
+			return nil, err
+		}
+		opts = append(opts, grpc.Creds(credentials.NewTLS(serverConfig)))
+	}
+
+	return opts, nil
+}
+
+func (d *RPCFactory) GetInternodeClientTlsConfig() (*tls.Config, error) {
+	if d.tlsFactory != nil {
+		return d.tlsFactory.GenerateInternodeClientConfig()
+	}
+
+	return nil, nil
+}
+
+// GetGRPCListener returns cached dispatcher for gRPC inbound or creates one
+func (d *RPCFactory) GetGRPCListener() net.Listener {
```

Comment:
Longer term todo: these older ```rpc.go``` functions should return error. This is why I don't consume ```Create*GRPCConnection``` in tests.